podvm: allow setting custom agent-policy

snir911 opened the pull request:
by having custom.rego file set as agent policy CM

kubectl create configmap agent-policy --from-file=<path/to/custom.rego> -n openshift-sandboxed-containers-operator

also use relative links is it essential when passed to packer

**do not merge**

**I failed to test yet due to (probably) unaltered issues**

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -196,11 +196,18 @@ function prepare_source_code() {
-        ln -sf "${podvm_dir}"/files/etc/kata-opa/coco-default-policy.rego  "${podvm_dir}"/files/etc/kata-opa/default-policy.rego
+        ln -sf coco-default-policy.rego  "${podvm_dir}"/files/etc/kata-opa/default-policy.rego
     fi
+    echo "\nCurrent Agent Policy:\n" && cat "${podvm_dir}"/files/etc/kata-opa/default-policy.rego
```
Hmm afaik echo will not expand the escape sequences. `printf` needed ?

bpradipt replied:
While you are touching this file, it might be useful to split the sed command into multiline for easier readability. for example

```
  sed 's/default SetPolicyRequest := true/default SetPolicyRequest := false/;
   s/default ExecProcessRequest := true/default ExecProcessRequest := false/' \
        "${podvm_dir}"/files/etc/kata-opa/default-policy.rego > \
        "${podvm_dir}"/files/etc/kata-opa/coco-default-policy.rego
```
Btw on second thoughts should SetPolicyRequest be set to true for now ?

bpradipt replied:
So as default coco policy, we can disable `ReadStreamRequest` and `ExecProcessRequest` to start with and provide SetPolicy examples to enable/disable apis for trying out different scenario
Also we'll document (blog) the (wip) eventual solution with initData and measured disk. 
Sounds ok ?

gkurz replied:
> Btw on second thoughts should SetPolicyRequest be set to true for now ?

Newbie question : what does SetPolicyResquest do ?

> So as default coco policy, we can disable `ReadStreamRequest` and `ExecProcessRequest` to start with and provide SetPolicy examples to enable/disable apis for trying out different scenario

IIUC this is calling for another PR that re-enables `SetPolicyRequest` and disables `ReadStreamRequest`, right ?

bpradipt replied:
A new PR or at least a new commit explaining the rationale.

bpradipt commented:
Tested this successfully by building a new image
```
oc exec -it hello-b -- sh
error: Internal error occurred: error executing command in container: cannot enter container 6d9c0d50f7640003baed372f0804ada60e8f04ce8222a0de8ab6b6f48de6411c, with err rpc error: code = PermissionDenied desc = "ExecProcessRequest is blocked by policy: ": unknown
```

snir911 committed:
podvm: fix agent policy

by using relative path
and disable ReadStreamRequest and ExecProcessRequest for CoCo by default

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

snir911 committed:
podvm: allow setting custom agent-policy

by setting base64 encoded policy file string to the AGENT_POLICY key in the
<azure/aws-podvm>-image-cm ConfigMap

$ ENCODED_POLICY=$(cat allow-all-except-exec-process.rego | base64 -w 0)
$ kubectl patch cm aws-podvm-image-cm -p "{\"data\":{\"AGENT_POLICY\":\"${ENCODED_POLICY}\"}}" -n openshift-sandboxed-containers-operator

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -197,7 +197,12 @@ function prepare_source_code() {
 
     # links must be relative
-    if [[ "$CONFIDENTIAL_COMPUTE_ENABLED" == "yes" ]]; then
+    if [[ -e /tmp/kata-opa/custom.rego ]] ; then
```
How about adding a folder in podvm with sample policy file and a README describing the details on how policy is handled in OSC (via the configmap etc) along with examples on using the custom policy  as well as another example of setting policy via the annotation?

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -196,11 +196,14 @@ function prepare_source_code() {
+            s/default ExecProcessRequest := true/default ExecProcessRequest := false/' \
             "${podvm_dir}"/files/etc/kata-opa/default-policy.rego > "${podvm_dir}"/files/etc/kata-opa/coco-default-policy.rego
-        ln -sf "${podvm_dir}"/files/etc/kata-opa/coco-default-policy.rego  "${podvm_dir}"/files/etc/kata-opa/default-policy.rego
+        ln -sf coco-default-policy.rego "${podvm_dir}"/files/etc/kata-opa/default-policy.rego
```
Hmm is the above correct?
the coco-default-policy.rego is in `"${podvm_dir}"/files/etc/kata-opa/`

gkurz replied:
> Hmm is the above correct? the coco-default-policy.rego is in `"${podvm_dir}"/files/etc/kata-opa/`

Yes it is because both files are in the same directory. When the VFS will try to access the right most component of `"${podvm_dir}"/files/etc/kata-opa/default-policy.rego`  from `"${podvm_dir}"/files/etc/kata-opa/`, it will go to the exact content of the symlink, i.e. `coco-default-policy.rego` which exists in the directory.

The first argument to `ln -s` is basically arbitrary text that the VFS is expected to use when resolving the symlink. The `symlink(2)` syscall doesn't even require the target to exist.

snir911 committed:
podvm: add agent-policy README.md

and example files

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

snir911 committed:
podvm: disable set -x by default

to avoid unwanted printings

Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

gkurz commented on config/peerpods/podvm/lib.sh:
```diff
@@ -206,13 +206,22 @@ function prepare_source_code() {
+    if [[ "${AGENT_POLICY}" ]]; then
+        echo "Custom agent policy is being set through the AGENT_POLICY value"
+        echo ${AGENT_POLICY} | base64 -d > "${podvm_dir}"/files/etc/kata-opa/custom.rego
+        if [[ $? == 0 ]] && grep -q "agent_policy" "${podvm_dir}"/files/etc/kata-opa/custom.rego; then # checks policy validity
```
Isn't there a better way to validate the syntax than grepping ? Couldn't https://www.openpolicyagent.org/docs/latest/cli/#opa-check be used here ?

snir911 replied:
right, my intention wasn't to actually validate the format rather to validate the output isn't ruined because of missing couple of characters when copied the encoded string.
This opa check is indeed an option, it will require to install the binary in the container, can certainly be done at future point.

gkurz reviewed (approved):
Grep is really a poor man's tool to validate user input but we can start with this and improve later. Rest looks good to me.

Thanks @snir911 !

bpradipt commented on config/peerpods/podvm/agent-policy/README.md:
```diff
@@ -0,0 +1,53 @@
+file (e.g., [allow-all-except-exec-process.rego](allow-all-except-exec-process.rego)) in base64 format
+and set it to the `io.katacontainers.config.agent.policy` annotation.
+
+**note:** annotation policy will override any previous policy (as long `SetPolicyRequest` is allowed)
```
```suggestion
**note:** annotation policy will override any previous policy (as long as `SetPolicyRequest` is allowed)
```

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -205,11 +205,23 @@ function prepare_source_code() {
+        if [[ $? == 0 ]] && grep -q "agent_policy" "${podvm_dir}"/files/etc/kata-opa/custom.rego; then # checks policy validity
+            ln -sf custom.rego  "${podvm_dir}"/files/etc/kata-opa/default-policy.rego
+        else
+            error_exit "Invalid AGENT_POLICY value set, expected 64 based valid agent policy, got: \"${AGENT_POLICY}\""
```
```suggestion
            error_exit "Invalid AGENT_POLICY value set, expected base64 encoded valid agent policy, got: \"${AGENT_POLICY}\""
```

bpradipt commented on config/peerpods/podvm/lib.sh:
```diff
@@ -205,11 +205,23 @@ function prepare_source_code() {
+        else
+            error_exit "Invalid AGENT_POLICY value set, expected 64 based valid agent policy, got: \"${AGENT_POLICY}\""
+	fi
+    elif [[ "$CONFIDENTIAL_COMPUTE_ENABLED" == "yes" ]]; then
```
We don't want to allow custom policy when CONFIDENTIAL_COMPUTE_ENABLED is yes?

snir911 replied:
we don't? i think that if user is asking for certain policy explicitly we should accept it (it may be more restricted than the default)

bpradipt replied:
Ok, custom agent policy gets priority, which is fair imho. Thanks

bpradipt reviewed (approved):
/lgtm
Thanks @snir911
